File: student_management_app/lib/models/student.dart
import 'department.dart';

class Student {
  const Student({
    required this.id,
    required this.studentId,
    required this.fullName,
    required this.email,
    required this.department,
    this.gender,
    this.phone,
    this.year,
  });

  final String id;
  final String studentId;
  final String fullName;
  final String email;
  final Department department;
  final String? gender;
  final String? phone;
  final int? year;

  factory Student.fromJson(Map<String, dynamic> json) {
    final rawDepartment = json['department'];
    final department = rawDepartment is Map<String, dynamic>
        ? Department.fromJson(rawDepartment)
        : Department(
            id: rawDepartment?.toString() ?? '',
            name: rawDepartment?.toString() ?? 'Unknown',
          );

    return Student(
      id: json['_id']?.toString() ?? '',
      studentId: json['studentId']?.toString() ?? '',
      fullName: json['fullName']?.toString() ?? '',
      email: json['email']?.toString() ?? '',
      department: department,
      gender: json['gender']?.toString(),
      phone: json['phone']?.toString(),
      year: json['year'] is num ? (json['year'] as num).toInt() : null,
    );
  }
}

File: backend/models/Student.js
const mongoose = require("mongoose");

const studentSchema = new mongoose.Schema(
  {
    studentId: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      maxlength: 30,
    },
    fullName: {
      type: String,
      required: true,
      trim: true,
      minlength: 2,
      maxlength: 100,
    },
    gender: {
      type: String,
      enum: ["Male", "Female"],
    },
    email: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      lowercase: true,
      match: [/^[^\s@]+@[^\s@]+\.[^\s@]+$/, "Please enter a valid email"],
    },
    phone: {
      type: String,
      trim: true,
      maxlength: 30,
    },
    department: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Department",
      required: true,
    },
    classGroup: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "ClassGroup",
      default: null,
    },
    year: {
      type: Number,
      min: 1,
      max: 6,
      validate: {
        validator: Number.isInteger,
        message: "Year must be a whole number",
      },
    },
  },
  {
    timestamps: true,
  },
);

studentSchema.index({ department: 1, fullName: 1 });
studentSchema.index({ classGroup: 1, fullName: 1 });

module.exports = mongoose.model("Student", studentSchema);
